syncany-lib/src/main/java/org/syncany/plugins/transfer/features/AsyncFeatureTransferManager.java

```diff
@@ -0,0 +1,198 @@
+		Class<AsyncFeatureExtension> asyncFeatureExtensionClass = (Class<AsyncFeatureExtension>) asyncAnnotation.extension();
+
+		try {
+			Constructor<?> constructor = ReflectionUtil.getMatchingConstructorForClass(asyncFeatureExtensionClass, originalTransferManagerClass);
+
+			if (constructor != null) {
+				return (AsyncFeatureExtension) constructor.newInstance(originalTransferManager);
+			}
+
+			return asyncFeatureExtensionClass.newInstance();
+		}
+		catch (InvocationTargetException | InstantiationException | IllegalAccessException | NullPointerException e) {
+			throw new RuntimeException("Cannot instantiate AsyncFeatureExtension (perhaps " + asyncFeatureExtensionClass + " does not exist?)", e);
+		}
+	}
+
+	@Override
+	public void connect() throws StorageException {
+		underlyingTransferManager.connect();
+	}
+
+	@Override
+	public void disconnect() throws StorageException {
+		underlyingTransferManager.disconnect();
+	}
+
+	@Override
+	public void init(final boolean createIfRequired) throws StorageException {
+		underlyingTransferManager.init(createIfRequired);
+	}
+
+	@Override
+	public void download(final RemoteFile remoteFile, final File localFile) throws StorageException {
+		underlyingTransferManager.download(remoteFile, localFile);
+	}
+
+	@Override
+	public void move(final RemoteFile sourceFile, final RemoteFile targetFile) throws StorageException {
+		underlyingTransferManager.move(sourceFile, targetFile);
+		waitForFile(targetFile);
+	}
+
+	@Override
+	public void upload(final File localFile, final RemoteFile remoteFile) throws StorageException {
+		underlyingTransferManager.upload(localFile, remoteFile);
+		waitForFile(remoteFile);
+	}
+
+	@Override
+	public boolean delete(final RemoteFile remoteFile) throws StorageException {
+		return underlyingTransferManager.delete(remoteFile);
+	}
+
+	@Override
+	public <T extends RemoteFile> Map<String, T> list(final Class<T> remoteFileClass) throws StorageException {
+		return underlyingTransferManager.list(remoteFileClass);
+	}
+
+	@Override
+	public String getRemoteFilePath(Class<? extends RemoteFile> remoteFileClass) {
+		return underlyingTransferManager.getRemoteFilePath(remoteFileClass);
+	}
+
+	@Override
+	public StorageTestResult test(boolean testCreateTarget) {
+		return underlyingTransferManager.test(testCreateTarget);
+	}
+
+	@Override
+	public boolean testTargetExists() throws StorageException {
+		return underlyingTransferManager.testTargetExists();
+	}
+
+	@Override
+	public boolean testTargetCanWrite() throws StorageException {
+		return underlyingTransferManager.testTargetCanWrite();
+	}
+
+	@Override
+	public boolean testTargetCanCreate() throws StorageException {
+		return underlyingTransferManager.testTargetCanCreate();
+	}
+
+	@Override
+	public boolean testRepoFileExists() throws StorageException {
+		return underlyingTransferManager.testRepoFileExists();
+	}
+
+	private void waitForFile(RemoteFile remoteFile) throws StorageException {
+		while (true) {
```

Comment:
Maybe you can use a `ListenableFuture` here? At least you would get rid of the busy wating. However, I think to make this happen, the `exists` method needs to be blocking. Here is a sketch:
```java
ListeningExecutorService service = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(10));
ListenableFuture<?> remoteFileExists = service.submit(new Callable<RemoteFile>() {
  public RemoteFile call() {
   if(Thread.isInterrupted()) throw Throwables.throw(new InterruptedException("Interrupt"));      
   asyncFeatureExtension.exists(remoteFile);
   return remoteFile; 
  }
});
Futures.addCallback(explosion, new FutureCallback<?>() {
  public void onSuccess(RemoteFile remoteFile) {
    // ...
  }
  public void onFailure(Throwable thrown) {
    //...
  }
});
remoteFileExists.get(1, SECOND);
```